### ISSUE TITLE
How to generate code with reusable Union type?

### ISSUE THREAD

**Issue by bohan-v** [apollographql/apollo-ios#2931]
### Question

Hello everyone, 

I want to ask a question regarding Union Type code generation. I notice the union type inherits parent's type name. If there are recursive manner data structure, the generated union type can't reuse.

The following demonstrate the issue.

**Example Schema**
```
union Content = Section | Text

type Text {
  text: String!
}

type Section {
  contents: [Content!]!
  id: ID!
  name: String!
}
```

**Example Fragment Definition**
```
fragment SectionFragment on Section {
  id
  contents {
    ... on Section {
      name
      contents {
        ... on Section {
          name
          contents {
            ... on Text {
              ...TextFragment
            }
          }
        }
        ... on Text {
          ...TextFragment
        }
      }
    }
    ... on Text {
      ...TextFragment
    }
  }
}

fragment TextFragment on Text {
  text
}

```

The generated code at lowest nested level namespace is `SectionFragment.Content.AsSection.Content.AsSection.Content`

❓ Is there a way to configure the Apollo CLI to generate code that reuses the "Content" Union type more efficiently? 

I know we should flatten the data structure. But I would like to know how to make Union type.
```[tasklist]
```

**Comment by AnthonyMDev** [comment]
Thanks for the question! This is definitely a pain point right now. Its going to take a lot of complicated logic in the codegen engine to be able to recognize when types can be shared and reused across contexts. But this is something we have planned to address eventually. See issue #2191 for tracking this request.

**Comment by janakmshah** [comment]
This was closed as `completed` @AnthonyMDev, but the Hoisted Types feature has not been added to the SDK yet, right? Should this issue's status be `open`?

**Comment by fnazarios** [comment]
+1. I've been rewriting many of the response models and not being able to use the code gen structs because of it

**Comment by AnthonyMDev** [comment]
We just closed this as a duplicate of #2191, since the work that needs to be done there will resolve this issue. It's still being tracked and is on our roadmap.
